File: app/src/main/java/com/capstone/project_niyakneyak/login/activity/RegisterActivity.java
package com.capstone.project_niyakneyak.login.activity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.capstone.project_niyakneyak.R;
import com.capstone.project_niyakneyak.data.user_model.UserAccount ;
import com.capstone.project_niyakneyak.main.activity.MainActivity;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

public class RegisterActivity extends AppCompatActivity {

    private FirebaseAuth mFirebaseAuth; //파이어베이스 인증
    private DatabaseReference mDatabaseRef; //실시간 데이터베이스, 서버에 연동시키는 객체
    private EditText mEtEmail, mEtPassword, mEtName, mEtPhoneNum; //회원가입 입력필드
    private Button mBtnRegister;// 회원가입 버튼

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // activity_register.xml을 바인딩합니다.
        setContentView(R.layout.activity_register);

        mFirebaseAuth = FirebaseAuth.getInstance();
        mDatabaseRef = FirebaseDatabase.getInstance().getReference("niyakneyak");

        mEtEmail = findViewById(R.id.et_email);
        mEtPassword = findViewById(R.id.et_password);
        mEtName = findViewById(R.id.et_name);
        mEtPhoneNum = findViewById(R.id.et_phoneNum);

        mBtnRegister = findViewById(R.id.btn_register); // Initialize mBtnRegister

        mBtnRegister.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String strEmail = mEtEmail.getText().toString();
                String strPwd = mEtPassword.getText().toString();

                mFirebaseAuth.createUserWithEmailAndPassword(strEmail, strPwd).addOnCompleteListener(RegisterActivity.this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {

                        if (task.isSuccessful()) {//회원가입 성공
                            FirebaseUser firebaseUser = mFirebaseAuth.getCurrentUser();//인증객체에서 현재의 유저를 가져옴
                            UserAccount account = new UserAccount();
                            account.setIdToken(firebaseUser.getUid());
                            account.setEmailId(firebaseUser.getEmail());
                            account.setPassword(strPwd);
                            account.setName(mEtName.getText().toString());
                            account.setPhoneNum(mEtPhoneNum.getText().toString());

                            //니약내약의 하위개념으로 정보를 넣음
                            //setValue를 통해 데이터베이스에 삽입
                            mDatabaseRef.child("UserAccount").child(firebaseUser.getUid()).setValue(account);
                            //가입이 이루어졌을 때, 가입 정보를 데이터베이스에 저장

                            Toast.makeText(RegisterActivity.this, "회원가입에 성공하셨습니다.", Toast.LENGTH_SHORT).show();

                            Intent intent = new Intent(RegisterActivity.this, MainActivity.class);
                            startActivity(intent);
                        } else {
                            Toast.makeText(RegisterActivity.this, "회원가입에 실패하셨습니다.", Toast.LENGTH_SHORT).show();
                            task.getException().getMessage();
                            Log.d("RegisterActivity", "Register Failed");
                        }
                    }
                });

            }
        });
    }
}

File: app/src/main/java/com/capstone/project_niyakneyak/login/activity/RegisterActivity.kt
package com.capstone.project_niyakneyak.login.activity

import android.app.DatePickerDialog
import android.os.Bundle
import android.util.Log
import android.widget.RadioButton
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.capstone.project_niyakneyak.data.user_model.UserAccount
import com.capstone.project_niyakneyak.databinding.ActivityRegisterBinding
import com.google.firebase.Firebase
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.auth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.firestore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import java.util.Calendar
import kotlinx.coroutines.*

class RegisterActivity : AppCompatActivity() {
    private lateinit var binding: ActivityRegisterBinding
    private lateinit var auth: FirebaseAuth
    private lateinit var firestore: FirebaseFirestore
    private val uiScope = CoroutineScope(Dispatchers.Main)

    companion object {
        private const val TAG = "REGISTER_ACTIVITY"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityRegisterBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Initialize Firebase Auth and Firestore
        auth = Firebase.auth
        firestore = Firebase.firestore

        // Set up click listener for the birth date TextView
        binding.tvBirth.setOnClickListener {
            showDatePickerDialog()
        }

        binding.btnRegister.setOnClickListener {
            val strEmail = binding.etEmail.text.toString().trim()
            val strPwd = binding.etPassword.text.toString().trim()

            if (strEmail.isNotEmpty() && strPwd.isNotEmpty()) {
                val selectedGenderId = binding.genderRadioGroup.checkedRadioButtonId

                val selectedGender = if (selectedGenderId != -1) {
                    findViewById<RadioButton>(selectedGenderId).text.toString()
                } else {
                    "성별 선택 안됨"
                }

                // The birthDate value is directly taken from the TextView
                val birthDate = binding.tvBirth.text.toString()

                auth.createUserWithEmailAndPassword(strEmail, strPwd)
                    .addOnSuccessListener(this@RegisterActivity) {
                        val firebaseUser = auth.currentUser
                        val account = UserAccount(
                            firebaseUser?.uid,
                            firebaseUser?.email,
                            strPwd,
                            binding.etName.text.toString(),
                            birthDate,
                            selectedGender,
                            binding.etPhoneNum.text.toString(),
                            age = calculateAge(birthDate)
                        )

                        firestore.collection(UserAccount.COLLECTION_ID).document(account.idToken!!).set(account)
                            .addOnSuccessListener {
                                setResult(RESULT_OK)
                                finish()
                            }
                            .addOnFailureListener {
                                Toast.makeText(this@RegisterActivity, "회원가입에 실패하셨습니다.", Toast.LENGTH_SHORT).show()
                                Log.w(TAG, "Terrible Error Occurred!: $it")
                            }
                    }.addOnFailureListener {
                        Toast.makeText(this@RegisterActivity, "회원가입에 실패하셨습니다.", Toast.LENGTH_SHORT).show()
                        Log.w(TAG, "Terrible Error Occurred!: $it")
                    }
            } else {
                Toast.makeText(this@RegisterActivity, "이메일과 비밀번호를 입력해주세요.", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun showDatePickerDialog() {
        val c = Calendar.getInstance()
        val year = c.get(Calendar.YEAR)
        val month = c.get(Calendar.MONTH)
        val day = c.get(Calendar.DAY_OF_MONTH)

        val dpd = DatePickerDialog(this, DatePickerDialog.OnDateSetListener { _, selectedYear, selectedMonth, selectedDay ->
            // Format and display the date in the TextView
            val dateStr = String.format("%d-%02d-%02d", selectedYear, selectedMonth + 1, selectedDay)
            binding.tvBirth.text = dateStr

            // Update age when a date is selected
            updateAge(dateStr)  // Call this function here to update the age
        }, year, month, day)

        dpd.show()
    }

    private fun updateAge(birthDate: String) {
        uiScope.launch {
            val age = withContext(Dispatchers.Default) {
                calculateAge(birthDate)  // 백그라운드 스레드에서 나이 계산
            }
            binding.tvAge.text = "만 $age 세"  // UI 스레드에서 TextView 업데이트
        }
    }
    private fun calculateAge(birthDateString: String): Int {
        val parts = birthDateString.split("-")
        if (parts.size < 3) return 0  // 날짜 형식이 잘못되었을 경우 안전하게 처리

        val birthYear = parts[0].toInt()
        val birthMonth = parts[1].toInt()
        val birthDay = parts[2].toInt()

        val today = Calendar.getInstance()
        val currentYear = today.get(Calendar.YEAR)
        val currentMonth = today.get(Calendar.MONTH) + 1  // Calendar.MONTH는 0부터 시작하므로 1을 더해줌
        val currentDay = today.get(Calendar.DAY_OF_MONTH)

        var age = currentYear - birthYear

        // 만약 생일이 아직 지나지 않았다면 나이에서 1을 빼준다
        if (birthMonth > currentMonth || (birthMonth == currentMonth && birthDay > currentDay)) {
            age -= 1
        }

        return age
    }

    override fun onDestroy() {
        super.onDestroy()
        uiScope.cancel()  // 액티비티가 종료될 때 코루틴 취소
    }

}
